### PR TITLE
[WIP] Fixes #2188: Add unit tests for create_challenge_using_zip_file API.

### LINKED ISSUE #2188: Add unit tests for create challenge API
## Current Scenario
There are no unit tests for create challenge using zip [API](https://github.com/Cloud-CV/EvalAI/blob/c156f3d54c95861c4ac4eaef1d2e3419bab83a7a/apps/challenges/views.py#L488) 

## Deliverables
- [ ] Add unit tests for the API

### PR DESCRIPTION
**Pull request by KhalidRmb**
Fixes #2188

### CODE REVIEW

File: tests/unit/challenges/test_views.py [hunk 1]
```diff
@@ -2510,13 +2514,31 @@ def setUp(self):
...
             "challenges:create_challenge_using_zip_file",
             kwargs={"challenge_host_team_pk": self.challenge_host_team.pk},
         )
+
+        self.BASE_TEMP_LOCATION = tempfile.mkdtemp()
+        self.base_path = join(settings.BASE_DIR, 'tests', 'unit', 'challenges', 'data')
+
+        self.path_to_annotation = join(self.base_path, 'annotations')
+        self.path_to_original_yaml_file = join(self.base_path, 'challenge_config.yaml')
+        self.path_to_altered_yaml = join(self.BASE_TEMP_LOCATION, "altered_yaml_file.yaml")
```
Comment (RishabhJain2018) [inline tests/unit/challenges/test_views.py]: `self.path_to_altered_yaml` -> `self.altered_challenge_config_path`

File: tests/unit/challenges/test_views.py [hunk 2]
```diff
@@ -2510,13 +2514,31 @@ def setUp(self):
...
+        self.path_to_sample_file = join(self.BASE_TEMP_LOCATION, 'sample.txt')
```
Comment (RishabhJain2018) [inline tests/unit/challenges/test_views.py]: `self.path_to_sample_file` -> `self.challenge_config_in_txt_format_path`

File: tests/unit/challenges/test_views.py [hunk 3]
```diff
@@ -2510,13 +2514,31 @@ def setUp(self):
...
+        self.path_to_eval_script_zip = join(self.base_path, 'evaluation_script.zip')
```
Comment (RishabhJain2018) [inline tests/unit/challenges/test_views.py]: `self.path_to_eval_script_zip` -> `self.evaluation_script_file_path`

File: tests/unit/challenges/test_views.py [hunk 4]
```diff
@@ -2510,13 +2514,31 @@ def setUp(self):
...
+
+        self.yaml_file = open(join(self.base_path, 'challenge_config.yaml'))
+        self.yaml_dict = yaml.safe_load(self.yaml_file)
+        self.copy_dict = copy.deepcopy(self.yaml_dict)
```
Comment (RishabhJain2018) [inline tests/unit/challenges/test_views.py]: Why are you using `deepcopy` here?

File: tests/unit/challenges/test_views.py [hunk 5]
```diff
@@ -2510,13 +2514,31 @@ def setUp(self):
...
+
+        self.alt_file = open(self.path_to_altered_yaml, 'w+')
+        self.alt_file.write("Sample")
```
Comment (RishabhJain2018) [inline tests/unit/challenges/test_views.py]: This is not a valid yaml? Please create an invalid yaml alongside with valid yaml.

File: tests/unit/challenges/test_views.py [hunk 6]
```diff
@@ -2622,6 +2626,176 @@ def test_create_challenge_using_zip_file_success(self):
         self.assertEqual(Leaderboard.objects.count(), 2)
         self.assertEqual(ChallengePhaseSplit.objects.count(), 2)
 
+    def test_create_challenge_using_zip_file_when_not_a_zip_file(self):
```
Comment (RishabhJain2018) [inline tests/unit/challenges/test_views.py]: `test_create_challenge_using_zip_file_when_not_a_zip_file` -> `test_create_challenge_using_zip_file_when_uploaded_file_isnt_a_zip_file`

File: tests/unit/challenges/test_views.py [hunk 7]
```diff
@@ -2622,6 +2626,176 @@ def test_create_challenge_using_zip_file_success(self):
...
+        samplefile = open(self.path_to_sample_file, 'w+')
+        samplefile.write("Test!")
```
Comment (RishabhJain2018) [inline tests/unit/challenges/test_views.py]: Please change the content to - `Sample challenge configuration`

File: tests/unit/challenges/test_views.py [hunk 8]
```diff
@@ -2622,6 +2626,176 @@ def test_create_challenge_using_zip_file_success(self):
...
+
+    # Helper function for tests below.
+    def create_challenge_test(self):
+        try:
+            exec(self.element_to_delete)
+            a = open(self.path_to_altered_yaml, 'r+')
+            yaml.dump(self.copy_dict, a, default_flow_style=False)
+        except KeyError: # To catch the case when no element is to be deleted- i.e, empty string is passed as key.
+            pass
+
```
Comment (RishabhJain2018) [inline tests/unit/challenges/test_views.py]: What is the use of the above try/except?

File: tests/unit/challenges/test_views.py [hunk 9]
```diff
@@ -2622,6 +2626,176 @@ def test_create_challenge_using_zip_file_success(self):
...
+    def test_create_challenge_using_zip_file_when_eval_script_key_missing(self):
+        self.filenames = [self.path_to_altered_yaml, self.path_to_eval_script_zip]
+        self.message = ('There is no key for evaluation script in YAML file. '
+                        'Please add it and then try again!')
+        self.element_to_delete = "del self.copy_dict['evaluation_script']"
+        self.status_code = status.HTTP_406_NOT_ACCEPTABLE
+        self.create_challenge_test()
+
+    #@create_challenge_test
```
Comment (RishabhJain2018) [inline tests/unit/challenges/test_views.py]: Please remove the comments

File: tests/unit/challenges/test_views.py [hunk 10]
```diff
@@ -2622,6 +2626,176 @@ def test_create_challenge_using_zip_file_success(self):
...
+    def test_create_challenge_using_zip_file_when_no_eval_script_present(self):
+        self.filenames = [self.path_to_altered_yaml]
+        self.message = ('No evaluation script is present in the zip file. '
+                        'Please add it and then try again!')
+        self.element_to_delete = "del self.copy_dict['']"
+        self.status_code = status.HTTP_406_NOT_ACCEPTABLE
+        self.create_challenge_test()
+
+    #def test_create_challenge_using_zip_file_when_yaml_syntax_error(self):
```
Comment (RishabhJain2018) [inline tests/unit/challenges/test_views.py]: Please remove the comments.

File: tests/unit/challenges/test_views.py [hunk 11]
```diff
@@ -2622,6 +2626,176 @@ def test_create_challenge_using_zip_file_success(self):
...
+        self.create_challenge_test()
+
+    def test_create_challenge_using_zip_file_when_no_key_for_eval_details(self):
+        self.filenames = [self.path_to_altered_yaml, self.path_to_eval_script_zip]
+        self.message = ('There is no key for evalutaion details. '
+                        'Please add it and then try again!')
+        self.element_to_delete = "del self.copy_dict['evaluation_details']"
+        self.status_code = status.HTTP_406_NOT_ACCEPTABLE
+
+    #@create_challenge_test
```
Comment (RishabhJain2018) [inline tests/unit/challenges/test_views.py]: Please remove the comments

File: tests/unit/challenges/test_views.py [hunk 12]
```diff
@@ -2622,6 +2626,176 @@ def test_create_challenge_using_zip_file_success(self):
...
+        self.filenames = [self.path_to_altered_yaml, self.path_to_eval_script_zip]
+        self.message = ('No test annotation file found in zip file'
+                        'for challenge phase \'{}\'. Please add it and '
+                        ' then try again!'.format(yaml_dict[challenge_phases][1][name]))
+        self.element_to_delete = "del self.copy_dict['']"
+        self.status_code = status.HTTP_406_NOT_ACCEPTABLE
+
+        shutil.rmtree(BASE_TEMP_LOCATION)
+
+    #Once PR is Merged:
```
Comment (RishabhJain2018) [inline tests/unit/challenges/test_views.py]: Please remove the comments.

**Comment by RishabhJain2018** [comment]
Also, please fix the travis build.

**Comment by KhalidRmb** [comment]
@RishabhJain2018 I'll make all the changes above. This is still a WIP, so I hadn't made the changes, and there are more tests to add. I'll do it soon.

**Pull request closed without merging**